### PR TITLE
feature/limit-order-book

### PR DESCRIPTION
**Pull request by phchan9**
Spec is https://www.notion.so/perp/Limit-Order-Spec-5394b52bf33c463e98ee3ead6c81ff9b

### PR Reminders
---

1. **Contract**: make sure the code follows our convention, ex: naming, explicit returns, etc.; if uncertain, discuss with others

2. **Test**: make sure tests can cover most normal and edge cases; if not, open follow-up tickets. Also, look out for failed tests and fix them!

3. **CHANGELOG.md**: update when external interfaces are changed

4. Workflow: 
    - Github: assign the pr to yourself; if pairing, can merge directly; else, assign someone to help review
    - Asana: assign the corresponding ticket to yourself and leave the pr link on it for easier follow-ups
    - Discord: if someone is mentioned in this pr, tag on Discord

### CODE REVIEW

File: test/limitOrderBook/limitOrderBook.fillOrder.test.ts [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+            isExactInput: true,
+            amount: parseEther("300"),
+            oppositeAmountBound: parseEther("0.1"),
+            deadline: ethers.constants.MaxUint256,
+            reduceOnly: false,
+        }
+
+        await expect(limitOrderBook.connect(trader).cancelLimitOrder(limitOrder)).to.be.revertedWith("LOB_OSMBS")
+    })
+})
```
Comment (vinta) [inline test/limitOrderBook/limitOrderBook.fillOrder.test.ts:327]: Might be better to have some failed cases for deadline or price/amount is not matched.

**Review by vinta — CHANGES_REQUESTED** [review]

File: contracts/limitOrder/LimitOrderBook.sol [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    /// @inheritdoc ILimitOrderBook
+    function cancelLimitOrder(LimitOrder memory order) external override {
+        // LOB_OSMBS: Order's Signer Must Be Sender
+        require(_msgSender() == order.trader, "LOB_OSMBS");
+
+        // we didn't require `signature` as input like fillLimitOrder(),
+        // so trader can actually cancel an order that is not existed
+        bytes32 orderHash = getOrderHash(order);
```
Comment (vinta) [inline contracts/limitOrder/LimitOrderBook.sol:138]: I'm not sure about this, should we change to `cancelLimitOrder(LimitOrder memory order, bytes memory signature)`?

File: contracts/interface/ILimitOrderFeeVault.sol [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    /// @param rewardToken The new address of rewardToken
+    event RewardTokenChanged(address rewardToken);
+
+    /// @notice Emitted when limitOrderBook is changed
+    /// @param limitOrderBook The new address of limitOrderBook
+    event LimitOrderBookChanged(address limitOrderBook);
+
+    /// @notice Emitted when feeAmount is changed
+    /// @param feeAmount The new feeAmount
+    event FeeAmountChanged(uint256 feeAmount);
```
Comment (vinta) [inline contracts/interface/ILimitOrderFeeVault.sol:15]: Should we replace `fee` with `reward`?

File: contracts/limitOrder/LimitOrderBook.sol [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+    function getOrderHash(LimitOrder memory order) public view returns (bytes32) {
+        return _hashTypedDataV4(keccak256(abi.encode(LIMIT_ORDER_TYPEHASH, order)));
+    }
+
+    function verifySigner(LimitOrder memory order, bytes memory signature) public view returns (address) {
+        bytes32 orderHash = getOrderHash(order);
+        address signer = ECDSAUpgradeable.recover(orderHash, signature);
+
+        // LOB_SNET: signer is not trader
```
Comment (tienshaoku) [inline contracts/limitOrder/LimitOrderBook.sol:165]: Comment not aligned with the real error message!

File: contracts/limitOrder/LimitOrderFeeVault.sol [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    function setFeeAmount(uint256 feeAmountArg) external onlyOwner {
+        // LOFV_FAMBGT0: FeeAmount Must Be Greater Than 0
+        require(feeAmountArg > 0, "LOFV_FAMBGT0");
+        feeAmount = feeAmountArg;
+        emit FeeAmountChanged(feeAmountArg);
+    }
+
+    function disburse(address keeper, uint256 orderValue)
```
Comment (paco0x) [inline contracts/limitOrder/LimitOrderFeeVault.sol:68]: `orderValue` is unused

File: contracts/limitOrder/LimitOrderBook.sol (hunk 2 above)
Comment (KimiWu123) [reply contracts/limitOrder/LimitOrderBook.sol:147]: In the first place, I agree with you but after thinking a while, I think we should keep what we have now. The reason why is traders will submit their limit orders more carefully bcs they need to pay gas to cancel it. Might mitigate our keeper being attacked by spam orders.

File: contracts/interface/ILimitOrderFeeVault.sol (hunk 3 above)
Comment (KimiWu123) [reply contracts/interface/ILimitOrderFeeVault.sol:15]: agree, we should use `reward` instead.

File: contracts/limitOrder/LimitOrderBook.sol [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        // LOB_LOFVINC : LimitOrderFeeVault Is Not Contract
+        require(limitOrderFeeVaultArg.isContract(), "LOB_LOFVINC");
+        limitOrderFeeVault = limitOrderFeeVaultArg;
+    }
+
+    function setClearingHouse(address clearingHouseArg) external onlyOwner {
+        // LOB_CHINC: ClearingHouse Is Not a Contract
+        require(clearingHouseArg.isContract(), "LOB_CHINC");
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderBook.sol:68]: better to check `accountBalance` is not null. It might happen when `clearingHouse` is a contract but not our CH

File: contracts/limitOrder/LimitOrderBook.sol [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+        verifySigner(order, signature);
+
+        // TODO: support StopLimitOrder in the future
+        // LOB_OSLO: Only Support LimitOrder
+        require(order.orderType == ILimitOrderBook.OrderType.LimitOrder, "LOB_OSLO");
+
+        // LOB_OIFA: Order is filled already
+        require(_ordersStatus[orderHash] != ILimitOrderBook.OrderStatus.Filled, "LOB_OIFA");
+        // LOB_OIC: Order is cancelled
+        require(_ordersStatus[orderHash] != ILimitOrderBook.OrderStatus.Cancelled, "LOB_OIC");
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply contracts/limitOrder/LimitOrderBook.sol:97]: agree with Paco, the requirement here should be `unFilled` only.

File: contracts/limitOrder/LimitOrderBook.sol [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+            order.baseToken
+        );
+
+        if (order.reduceOnly) {
+            // LOB_NRO: Not ReduceOnly
+            require(
+                oldTakerPositionSize.mul(newTakerPositionSize) > 0 &&
+                    oldTakerPositionSize.abs() > newTakerPositionSize.abs(),
+                "LOB_NRO"
+            );
```
[the remark this answers is not in the document]
Comment (KimiWu123) [reply contracts/limitOrder/LimitOrderBook.sol:129]: we can replace this `require((oldTakerPositionSize < 0 != order.isBaseToQuote) && oldTakerPositionSize.abs() >= base)`. If we do so, we don't need to get `newTakerPositionSize` and can save an external call.

File: contracts/limitOrder/LimitOrderFeeVault.sol [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        external
+        override
+        onlyLimitOrderBook
+        nonReentrant
+        returns (uint256)
+    {
+        // TODO: be aware of decimal issue when we use different reward token
+
+        // LOFV_NEBTD: not enough balance to disburse
+        require(IERC20Upgradeable(rewardToken).balanceOf(address(this)) >= feeAmount, "LOFV_NEBTD");
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderFeeVault.sol:78]: question, do we really want to revert here? If so, it'll cause trader's order being reverted and
block keepers like us who are willing to execute the orders without feedback.

File: contracts/limitOrder/LimitOrderFeeVault.sol [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+        SafeERC20Upgradeable.safeTransfer(IERC20Upgradeable(rewardToken), keeper, feeAmount);
+
+        emit Disbursed(keeper, feeAmount);
+
+        return feeAmount;
+    }
+
+    // TODO: should we support multiple token to withdraw? or only support rewardToken?
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderFeeVault.sol:87]: I think we could support one token at the beginning. `disburse()` only supports one token as well. We should align the behaviors of all the functions.

File: contracts/limitOrder/LimitOrderFeeVault.sol [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    function withdraw(address token, uint256 amount) external override onlyOwner nonReentrant {
+        // LOFV_WTMBRT: Withdrawn Token Must Be RewardToken
+        require(token == rewardToken, "LOFV_WTMBRT");
+
+        address to = _msgSender();
+
+        // LOFV_NEBTW: not enough balance to withdraw
+        require(IERC20Upgradeable(token).balanceOf(address(this)) >= amount, "LOFV_NEBTW");
+
+        SafeERC20Upgradeable.safeTransfer(IERC20Upgradeable(token), to, amount);
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderFeeVault.sol:97]: if this function is designed for `owner` to withdraw, suggest we should hard code `to` to `owner`. It's safer if we accidentally remove `onlyOwner`.

File: test/limitOrderBook/limitOrderBook.fillOrder.test.ts [hunk 12]
```diff
@@ -0,0 +1,575 @@
...
+        await mintAndDeposit(fixture, trader, 1000)
+
+        await delegateApproval.connect(trader).approve([
+            {
+                delegate: limitOrderBook.address,
+                action: fixture.clearingHouseOpenPositionAction,
+            },
+        ])
+    })
+
```
Comment (KimiWu123) [inline test/limitOrderBook/limitOrderBook.fillOrder.test.ts:124]: we need one more case to specify the spec. what will happen if we short 0.1 position in this case? should be success or not?

File: contracts/limitOrder/LimitOrderBook.sol [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+        (uint256 base, uint256 quote) = IClearingHouse(clearingHouse).openPositionFor(
+            order.trader,
+            IClearingHouse.OpenPositionParams({
+                baseToken: order.baseToken,
+                isBaseToQuote: order.isBaseToQuote,
+                isExactInput: order.isExactInput,
+                amount: order.amount,
+                oppositeAmountBound: order.oppositeAmountBound,
+                deadline: order.deadline,
+                sqrtPriceLimitX96: 0,
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderBook.sol:113]: have we considered to leverage this parameter, `sqrtPriceLimitX96` in limit order? If traders  can set this value properly, their order could be partial filled instead of reverted when they have big orders.

File: contracts/limitOrder/LimitOrderBook.sol [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+                referralCode: ""
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderBook.sol:114]: Do we not allow referral program to use our limit order? Suggest we could support it if there is no strong preference.

File: test/limitOrderBook/limitOrderBook.fillOrder.test.ts [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+            roundIdWhenCreated: parseEther("0").toString(),
+            triggerPrice: parseEther("0").toString(),
+        }
+
+        await expect(await limitOrderBook.connect(trader).cancelLimitOrder(limitOrder)).to.emit(
+            limitOrderBook,
+            "LimitOrderCancelled",
+        )
+    })
+
```
Comment (KimiWu123) [inline test/limitOrderBook/limitOrderBook.fillOrder.test.ts:347]: what would happen if we cancel an order twice? We might need a test case to clarify the spec.

File: test/limitOrderBook/limitOrderBook.fillOrder.test.ts [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+        // sign limit order
+        const signature = await getSignature(fixture, limitOrder, trader)
+        const orderHash = await getOrderHash(fixture, limitOrder)
+
+        const tx = await limitOrderBook.connect(keeper).fillLimitOrder(limitOrder, signature, parseEther("0"))
+        await expect(tx)
+            .to.emit(limitOrderBook, "LimitOrderFilled")
+            .withArgs(trader.address, baseToken.address, orderHash, keeper.address, fixture.rewardAmount)
+
+        await expect(tx).to.emit(limitOrderFeeVault, "Disbursed").withArgs(keeper.address, fixture.rewardAmount)
```
Comment (KimiWu123) [inline test/limitOrderBook/limitOrderBook.fillOrder.test.ts:151]: can add one more check, check keeper's balance to make sure the keeper really get reward.

File: contracts/limitOrder/LimitOrderFeeVault.sol (hunk 9 above)
Comment (phchan9) [reply contracts/limitOrder/LimitOrderFeeVault.sol:78]: I guess we need to setup monitor(like grafana or oncall duty) to check if this balance is enough to disburse.

Besides, if we run out of perp token in this LimitOrderFeeVault, we still make this order filled but won't send any reward to users. Later we found this situation happened, we need to write an off-chain script to send reward to those keepers, which could take us more effort to handle it).

File: test/limitOrderFeeVault/limitOrderFeeVault.test.ts [hunk 17]
```diff
@@ -0,0 +1,264 @@
+// import { FakeContract } from "@defi-wonderland/smock"
```
[the remark this answers is not in the document]
Comment (phchan9) [reply test/limitOrderFeeVault/limitOrderFeeVault.test.ts:1]: we will add this one back with several cases

File: contracts/limitOrder/LimitOrderBook.sol (hunk 13 above)
Comment (phchan9) [reply contracts/limitOrder/LimitOrderBook.sol:113]: not sure how to use `sqrtPriceLimitX96` for partial filled but it might increase more complexity based on the current spec. We could discuss this more on gather or meetings and prefer not add this feature in the current release

Comment (KimiWu123) [reply contracts/limitOrder/LimitOrderBook.sol:111]: actually don't need to do anything from contract side, what we need to do is have this parameter that can be passed from outside. Just for flexibility.

File: contracts/limitOrder/LimitOrderBook.sol (hunk 6 above)
Comment (KimiWu123) [reply contracts/limitOrder/LimitOrderBook.sol:72]: @vinta the checking should be included in `setClearingHouse` as well, not only `initialize`

File: contracts/limitOrder/LimitOrderFeeVault.sol [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+
+        SafeERC20Upgradeable.safeTransfer(IERC20Upgradeable(rewardToken), keeper, feeAmount);
+
+        emit Disbursed(keeper, feeAmount);
+
+        return feeAmount;
+    }
+
+    function withdraw(uint256 amount) external override onlyOwner nonReentrant {
+        address owner = _msgSender();
```
[the remark this answers is not in the document]
Comment (vinta) [reply contracts/limitOrder/LimitOrderFeeVault.sol:81]: Oh, I see.

File: contracts/limitOrder/LimitOrderBook.sol (hunk 13 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderBook.sol:115]: Will we need to record how much is partial filled? I suppose there could be multiple filled events for an order if it's partial filled multiple time?

What will happen if keeper call `fillLimitOrder()` on a previously-partial-fiilled order?

Comment (KimiWu123) [reply contracts/limitOrder/LimitOrderBook.sol:115]: My idea is simpler, `sqrtPriceLimitX96` will be set by traders. So if they set this value, their order will be partial filled if it fits the criteria. The state of this order would become `Filled` which means no more actions for this order. 

It's all offchain behaviors, we don't need to decide how it will work now. We can support this parameter first and still set `0` in our service. Everything will work just like what we have now. And we could activate this feature in the future after we discussed the details.

File: contracts/limitOrder/LimitOrderFeeVault.sol (hunk 9 above)
Comment (vinta) [reply contracts/limitOrder/LimitOrderFeeVault.sol]: Besides, if one filled order's keeper got keeper reward, but another filled order's keeper doesn't, it would be confused since both transactions are successful.

File: test/limitOrderBook/limitOrderBook.fillOrder.test.ts [hunk 19]
```diff
@@ -0,0 +1,575 @@
...
+
+            expect(await accountBalance.getTakerPositionSize(trader.address, baseToken.address)).to.gte(
+                parseEther("0.05"),
+            )
+
+            expect(await accountBalance.getTakerOpenNotional(trader.address, baseToken.address)).to.be.gte(
+                parseEther("-160"),
+            )
+        })
+
```
[the remark this answers is not in the document]
Comment (vinta) [reply test/limitOrderBook/limitOrderBook.fillOrder.test.ts:293]: added `fill order when fully close`

**Comment by vinta** [comment]
@KimiWu123 Feedbacks are fixed. Please take a look when you have time.

File: contracts/limitOrder/LimitOrderFeeVault.sol [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    function withdraw(uint256 amount) external override onlyOwner nonReentrant {
+        address owner = owner();
+
+        // LOFV_NEBTW: Not Enough Balance To Withdraw
+        require(IERC20Upgradeable(rewardToken).balanceOf(address(this)) >= amount, "LOFV_NEBTW");
+
+        SafeERC20Upgradeable.safeTransfer(IERC20Upgradeable(rewardToken), owner, amount);
+
+        emit Withdrawn(owner, rewardToken, amount);
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderFeeVault.sol:89]: not quite sure `owner` is necessary here bcs it's an `onlyOwner` function

File: contracts/limitOrder/LimitOrderBook.sol [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+        require(limitOrderFeeVaultArg.isContract(), "LOB_LOFVINC");
+        limitOrderFeeVault = limitOrderFeeVaultArg;
+    }
+
+    function setClearingHouse(address clearingHouseArg) external onlyOwner {
+        // LOB_CHINC: ClearingHouse Is Not a Contract
+        require(clearingHouseArg.isContract(), "LOB_CHINC");
+        clearingHouse = clearingHouseArg;
+        accountBalance = IClearingHouse(clearingHouse).getAccountBalance();
+
```
Comment (KimiWu123) [inline contracts/limitOrder/LimitOrderBook.sol:73]: suggest still need `require(accountBalanceArg.isContract(), "LOB_ABINC");` here. Compare to `initialize()`, it's more important to check here bcs when we need to reset CH address, it implies this contract is running and providing service. So we better make sure everything is correct.

**Review by KimiWu123 — APPROVED** [review]

**Review by paco0x — APPROVED** [review]

**Comment by Wraecca** [comment]
@vinta  there's $10 minimum amount from the spec but i didn't see it, do we still have it? if no, is there anyway we can prevent attacker drain the LimitOrderRewardVault?
